service/frontend/nexus_incoming_service_registry.go

```diff
@@ -0,0 +1,406 @@
+	defer r.RUnlock()
+
+	service, ok := r.servicesByName[request.Name]
+	if !ok {
+		return nil, serviceerror.NewNotFound(fmt.Sprintf("frontend could not find cached service with name %v", request.Name))
+	}
+
+	return &operatorservice.GetNexusIncomingServiceResponse{Service: service}, nil
+}
+
+func (r *registry) ListNexusIncomingServices(
+	ctx context.Context,
+	request *operatorservice.ListNexusIncomingServicesRequest,
+) (*operatorservice.ListNexusIncomingServicesResponse, error) {
+	if !r.isTableVersionCurrent(ctx) {
+		r.invalidateCachedServices()
+		if err := r.fetchServices(ctx); err != nil {
+			return nil, fmt.Errorf("error loading nexus incoming services frontend cache: %w", err)
+		}
+	}
+
+	r.RLock()
+	defer r.RUnlock()
+
+	startIdx := 0
+	if request.NextPageToken != nil {
+		nextServiceID := string(request.NextPageToken)
+
+		startFound := false
+		startIdx, startFound = slices.BinarySearch(r.serviceIDsSorted, nextServiceID)
+
+		if !startFound {
+			return nil, serviceerror.NewFailedPrecondition("could not find service indicated by nexus list incoming services next page token")
+		}
+	}
+
+	endIdx := startIdx + int(request.PageSize)
+	if endIdx > len(r.serviceIDsSorted) {
+		endIdx = len(r.serviceIDsSorted)
+	}
+
+	var nextPageToken []byte
+	if endIdx < len(r.services) {
+		nextPageToken = []byte(r.serviceIDsSorted[endIdx])
+	}
+
+	return &operatorservice.ListNexusIncomingServicesResponse{
+		NextPageToken: nextPageToken,
+		Services:      r.services[startIdx:endIdx],
+	}, nil
+}
+
+func (r *registry) isTableVersionCurrent(ctx context.Context) bool {
+	actualTableVersion, err := r.persistence.GetNexusIncomingServicesTableVersion(ctx)
+	if err != nil {
+		return false
+	}
+
+	r.RLock()
+	defer r.RUnlock()
+	return r.tableVersion == actualTableVersion
+}
+
+func (r *registry) fetchServices(ctx context.Context) error {
+	var currentPageToken []byte
+	pagingWasReset := false // used to ensure write lock is only acquired once
+
+	for ctx.Err() == nil {
+		r.RLock()
+		requestTableVersion := r.tableVersion
+		r.RUnlock()
+
+		shouldWait := currentPageToken == nil && requestTableVersion != 0
+
+		resp, err := r.matchingClient.ListNexusIncomingServices(ctx, &matchingservice.ListNexusIncomingServicesRequest{
+			LastKnownTableVersion: requestTableVersion,
+			NextPageToken:         currentPageToken,
+			PageSize:              loadServicesPageSize,
+			Wait:                  shouldWait,
+		})
+		if err != nil {
+			if errors.Is(err, p.ErrNexusTableVersionConflict) {
+				// indicates table was updated during paging, so reset and start from the beginning
+				currentPageToken = nil
+				pagingWasReset = true
+				r.invalidateCachedServicesLocked()
+				continue
+			}
+			// for all other errors, fall back to reading directly from persistence
+			if currentPageToken == nil && !pagingWasReset {
+				r.Lock()
+			}
+			persistenceErr := r.loadServicesLocked(ctx)
+			if persistenceErr == nil {
+				r.setServiceDataReady(persistenceErr)
```

Comment:
You don't want to permanently fail this future. This should be a situation you can recover from.

Reply:
It doesn't permanently fail it, the future is only set if persistence returned no error. I'll change it to set it to `nil` to make that clearer.